// @sky-mavis/waypoint/dist/commonjs/headless/track/track.js
'use strict';

var index = require('../../index-Pma4qYcb.js');
var common_version = require('../../common/version.js');
var headless_error_client = require('../error/client.js');
var headless_error_server = require('../error/server.js');
var headless_utils_serviceUrl = require('../utils/service-url.js');
var v4 = require('../../v4-D8BIpzex.js');
var sha256 = require('../../sha256-DGm_MAwc.js');
var toBytes = require('../../toBytes-Dmnw03lt.js');
require('../proto/rpc.js');
require('../../file-D9-_A4DE.js');
require('../../enum-CbdizYms.js');
require('../../sha256-CLylCSfy.js');
require('../../utils-U8Z-__N5.js');
require('../../size-BDN9jZuU.js');
require('../../base-BPC6VLRS.js');

const CONTENT_TYPE = "application/json";
const TRACK_URL = "https://x.skymavis.com/track";
const AUTHORIZATION_PROD = "Basic ODZkMTNkMmYtYWFmYy00M2YyLWJhZDctNDI2NTBiYmJmZTJlOg==";
const AUTHORIZATION_STAG = "Basic ZDU1ODQyODYtOWIwYS00MzE3LWI3YjktOWRjOTQwNmFiMzJlOg==";
exports.HeadlessEventName = void 0;
(function (HeadlessEventName) {
  HeadlessEventName["backupShard"] = "backupShard";
  HeadlessEventName["decryptShard"] = "decryptShard";
  HeadlessEventName["keygen"] = "keygen";
  HeadlessEventName["personalSign"] = "personalSign";
  HeadlessEventName["signTypedData"] = "signTypedData";
  HeadlessEventName["sendLegacyTransaction"] = "sendLegacyTransaction";
  HeadlessEventName["endEIP1559Transaction"] = "endEIP1559Transaction";
  HeadlessEventName["sendSponsoredTransaction"] = "sendSponsoredTransaction";
})(exports.HeadlessEventName || (exports.HeadlessEventName = {}));
const track = (events, isProdEnv) => {
  const headers = new Headers({});
  headers.set("Authorization", isProdEnv ? AUTHORIZATION_PROD : AUTHORIZATION_STAG);
  headers.set("Content-Type", CONTENT_TYPE);
  headers.set("Accept", CONTENT_TYPE);
  const body = JSON.stringify({
    events
  });
  return fetch(TRACK_URL, {
    method: "POST",
    headers,
    body
  });
};
const TRACKING_OFFSET_KEY = "WAYPOINT.HEADLESS.TRACKING_OFFSET";
const getOffset = () => {
  const storageReady = typeof window !== "undefined" && "sessionStorage" in window;
  if (!storageReady) return 0;
  try {
    const currentOffset = parseInt(sessionStorage.getItem(TRACKING_OFFSET_KEY) ?? "0");
    const newOffset = currentOffset + 1;
    sessionStorage.setItem(TRACKING_OFFSET_KEY, newOffset.toString());
    return currentOffset;
  } catch (error) {
    return 0;
  }
};
const getUTCTime = () => {
  const date = new Date();
  const timestamp = `${date.getUTCFullYear()}-${date.getUTCMonth()}-${date.getUTCDate()} ${date.getUTCHours()}:${date.getUTCMinutes()}:${date.getUTCSeconds()}`;
  return timestamp;
};
const toErrorProperties = error => {
  if (error instanceof headless_error_client.HeadlessClientError) return {
    error_level: "error",
    error_type: "client_defined",
    error_name: error.name,
    error_code: error.code,
    error_message: error.shortMessage
  };
  if (error instanceof headless_error_server.ServerError) return {
    error_level: "error",
    error_type: "server",
    error_name: error.name,
    error_code: error.code,
    error_message: error.shortMessage
  };
  if (error instanceof Error) return {
    error_level: "error",
    error_type: "client_unknown",
    error_name: error.name,
    error_code: headless_error_client.HeadlessClientErrorCode.UnknownError,
    error_message: error.message
  };
  return {
    error_level: "error",
    error_type: "client_unknown",
    error_name: "UnknownError",
    error_code: headless_error_client.HeadlessClientErrorCode.UnknownError,
    error_message: "Unknown error"
  };
};
const createTracker = params => {
  const {
    event,
    waypointToken,
    wasmUrl = "",
    productionFactor = false
  } = params;
  const startTime = performance.now();
  const isProdEnv = headless_utils_serviceUrl.isProd(productionFactor);
  const _getCommonData = () => {
    const endTime = performance.now();
    const duration = endTime - startTime;
    const {
      iss = "",
      sub = "",
      sid = "",
      aud = [],
      client_id = ""
    } = index.jwtDecode(waypointToken);
    return {
      event,
      uuid: v4.v4(),
      offset: getOffset(),
      timestamp: getUTCTime(),
      session_id: sid,
      user_id: iss && sub ? sha256.sha256(toBytes.stringToHex(`${iss}:${sub}`)) : "",
      ref: "",
      commonActionProperties: {
        aud,
        iss,
        sub,
        client_id,
        origin: window?.location?.origin ?? "",
        user_agent: navigator?.userAgent ?? "",
        sdk_version: common_version.version,
        wasm_version: wasmUrl,
        duration
      }
    };
  };
  const trackOk = okProperties => {
    try {
      const {
        commonActionProperties,
        ...restCommonData
      } = _getCommonData();
      const {
        request,
        response
      } = okProperties;
      const trackData = {
        ...restCommonData,
        action: "ok",
        action_properties: {
          ...commonActionProperties,
          request,
          response
        }
      };
      const trackEvent = {
        type: "track",
        data: trackData
      };
      track([trackEvent], isProdEnv);
    } catch (error) {
      /* empty */
    }
  };
  const trackError = error => {
    try {
      const {
        commonActionProperties,
        ...restCommonData
      } = _getCommonData();
      const errorProperties = toErrorProperties(error);
      const trackData = {
        ...restCommonData,
        action: "error",
        action_properties: {
          ...commonActionProperties,
          ...errorProperties
        }
      };
      const trackEvent = {
        type: "track",
        data: trackData
      };
      track([trackEvent], isProdEnv);
    } catch (error) {
      /* empty */
    }
  };
  return {
    trackOk,
    trackError
  };
};

exports.createTracker = createTracker;
