// @sky-mavis/waypoint/dist/commonjs/headless/action/sign-typed-data.js
'use strict';

var headless_error_client = require('../error/client.js');
var headless_track_track = require('../track/track.js');
var headless_action_getAddress = require('./get-address.js');
var headless_action_helpers_prepareTypedData = require('../../prepare-typed-data-D54m5_ZF.js');
var headless_action_sign = require('./sign.js');
var isAddress = require('../../isAddress-Ajbo9xO8.js');
var isAddressEqual = require('../../isAddressEqual-QoNFma_v.js');
var recoverAddress = require('../../recoverAddress-C3o2uHFx.js');
require('../../index-Pma4qYcb.js');
require('../../common/version.js');
require('../error/server.js');
require('../proto/rpc.js');
require('../../file-D9-_A4DE.js');
require('../../enum-CbdizYms.js');
require('../utils/service-url.js');
require('../../v4-D8BIpzex.js');
require('../../sha256-DGm_MAwc.js');
require('../../sha256-CLylCSfy.js');
require('../../utils-U8Z-__N5.js');
require('../../size-BDN9jZuU.js');
require('../../toBytes-Dmnw03lt.js');
require('../../base-BPC6VLRS.js');
require('../../secp256k1-CzvWuxSM.js');
require('../utils/convertor.js');
require('../../publicKeyToAddress-7k7v6GXz.js');
require('../../keccak256-DkMFbAQi.js');
require('../../concat-BguDurHH.js');
require('../../slice-Ca-rzTkU.js');
require('../proto/sign.js');
require('../utils/signature.js');
require('./helpers/authenticate.js');
require('../proto/auth.js');
require('../utils/token.js');
require('../../to-binary-B8Z0Mr_g.js');
require('./helpers/get-sign-handler.js');
require('../wasm/create.js');
require('../wasm/instantiate.js');
require('./helpers/open-socket.js');
require('./helpers/send-round-data.js');
require('./helpers/trigger-sign.js');

async function recoverTypedDataAddress(parameters) {
  const {
    domain,
    message,
    primaryType,
    signature,
    types
  } = parameters;
  return recoverAddress.recoverAddress({
    hash: headless_action_helpers_prepareTypedData.hashTypedData({
      domain,
      message,
      primaryType,
      types
    }),
    signature
  });
}

/**
 * Verify that typed data was signed by the provided address.
 *
 * Note:  Only supports Externally Owned Accounts. Does not support Contract Accounts.
 *        It is highly recommended to use `publicClient.verifyTypedData` instead to ensure
 *        wallet interoperability.
 *
 * - Docs {@link https://viem.sh/docs/utilities/verifyTypedData}
 *
 * @param parameters - {@link VerifyTypedDataParameters}
 * @returns Whether or not the signature is valid. {@link VerifyTypedDataReturnType}
 */
async function verifyTypedData(parameters) {
  const {
    address,
    domain,
    message,
    primaryType,
    signature,
    types
  } = parameters;
  return isAddressEqual.isAddressEqual(isAddress.getAddress(address), await recoverTypedDataAddress({
    domain,
    message,
    primaryType,
    signature,
    types
  }));
}

const signTypedData = async params => {
  const tracker = headless_track_track.createTracker({
    event: headless_track_track.HeadlessEventName.signTypedData,
    waypointToken: params.waypointToken,
    productionFactor: params.wsUrl,
    wasmUrl: params.wasmUrl
  });
  try {
    const {
      typedData,
      ...restParams
    } = params;
    const address = headless_action_getAddress.getAddressFromShard(params.clientShard);
    const rawMessage = headless_action_helpers_prepareTypedData.prepareTypedData(typedData);
    const signature = await headless_action_sign._sign({
      ...restParams,
      rawMessage: rawMessage
    });
    const isValid = await verifyTypedData({
      ...typedData,
      address: address,
      signature
    });
    if (!isValid) throw new headless_error_client.HeadlessClientError({
      cause: undefined,
      code: headless_error_client.HeadlessClientErrorCode.InvalidSignatureError,
      message: `Unable to verify the signature="${signature}" with the given address="${address}".`
    });
    tracker.trackOk({
      request: {
        typedData
      }
    });
    return signature;
  } catch (error) {
    tracker.trackError(error);
    throw error;
  }
};

exports.signTypedData = signTypedData;
