// @sky-mavis/waypoint/dist/commonjs/saigon-DYBomzZQ.js
'use strict';

function defineChain(chain) {
  return {
    formatters: undefined,
    fees: undefined,
    serializers: undefined,
    ...chain
  };
}

const ronin = /*#__PURE__*/defineChain({
  id: 2020,
  name: 'Ronin',
  nativeCurrency: {
    name: 'RON',
    symbol: 'RON',
    decimals: 18
  },
  rpcUrls: {
    default: {
      http: ['https://api.roninchain.com/rpc']
    }
  },
  blockExplorers: {
    default: {
      name: 'Ronin Explorer',
      url: 'https://app.roninchain.com'
    }
  },
  contracts: {
    multicall3: {
      address: '0xca11bde05977b3631167028862be2a173976ca11',
      blockCreated: 26023535
    }
  }
});

const saigon = /*#__PURE__*/defineChain({
  id: 2021,
  name: 'Saigon Testnet',
  nativeCurrency: {
    name: 'RON',
    symbol: 'RON',
    decimals: 18
  },
  rpcUrls: {
    default: {
      http: ['https://saigon-testnet.roninchain.com/rpc']
    }
  },
  blockExplorers: {
    default: {
      name: 'Saigon Explorer',
      url: 'https://saigon-app.roninchain.com'
    }
  },
  contracts: {
    multicall3: {
      address: '0xca11bde05977b3631167028862be2a173976ca11',
      blockCreated: 18736871
    }
  },
  testnet: true
});

exports.defineChain = defineChain;
exports.ronin = ronin;
exports.saigon = saigon;
