// @sky-mavis/waypoint/dist/commonjs/deposit/index.js
'use strict';

var common_communicate = require('../common/communicate.js');
var common_gate = require('../common/gate.js');
var common_popup = require('../common/popup.js');
require('../common/defer.js');
require('../common/waypoint-error.js');
require('../rpc-BJmuIVLE.js');
require('../base-BPC6VLRS.js');
require('../v4-D8BIpzex.js');
require('../common/Snackbar.js');

const DEPOSIT_POPUP_WIDTH = 500;
const DEPOSIT_POPUP_HEIGHT = 728;
class Deposit {
  clientId;
  waypointOrigin;
  redirectUri;
  environment;
  theme;
  communicateHelper;
  constructor(config) {
    const {
      waypointOrigin = common_gate.RONIN_WAYPOINT_ORIGIN_PROD,
      redirectUri = typeof window !== "undefined" ? window.location.origin : "",
      clientId,
      environment,
      theme
    } = config;
    this.waypointOrigin = waypointOrigin;
    this.clientId = clientId;
    this.redirectUri = redirectUri;
    this.environment = environment;
    this.theme = theme;
    this.communicateHelper = new common_communicate.CommunicateHelper(waypointOrigin);
  }
  start = async params => {
    const response = await this.communicateHelper.sendRequest(state => {
      const {
        email,
        walletAddress,
        fiatCurrency,
        cryptoCurrency,
        fiatAmount
      } = params ?? {};
      const query = {
        state,
        email,
        environment: this.environment,
        theme: this.theme,
        origin: this.redirectUri,
        redirect_uri: this.redirectUri,
        wallet_address: walletAddress,
        fiat_currency: fiatCurrency,
        crypto_currency: cryptoCurrency,
        fiat_amount: fiatAmount
      };
      const popupConfig = {
        width: DEPOSIT_POPUP_WIDTH,
        height: DEPOSIT_POPUP_HEIGHT
      };
      return common_popup.openPopup(`${this.waypointOrigin}/client/${this.clientId}/deposit`, query, popupConfig);
    });
    return {
      provider: response.provider,
      transactionHash: response.transaction_hash,
      fiatCurrency: response.fiat_currency,
      cryptoCurrency: response.crypto_currency,
      fiatAmount: response.fiat_amount,
      cryptoAmount: response.crypto_amount
    };
  };
}

exports.Deposit = Deposit;
