// @sky-mavis/waypoint/dist/commonjs/headless/utils/convertor.js
'use strict';

var toBytes = require('../../toBytes-Dmnw03lt.js');
require('../../base-BPC6VLRS.js');
require('../../size-BDN9jZuU.js');

// * STRING
// ! WHY: stringToBytes from "viem" do NOT working with base64
// * wasm & socket working with base64 string
const stringToBytes = value => {
  const charCode = Array.from(value, m => m.codePointAt(0));
  return Uint8Array.from(charCode);
};
const bytesToString = bytes => {
  return Array.from(bytes, b => String.fromCodePoint(b)).join("");
};
// * BASE64
const bytesToBase64 = bytes => btoa(bytesToString(bytes));
const base64ToBytes = base64 => stringToBytes(atob(base64));
const base64ToHex = base64 => toBytes.bytesToHex(base64ToBytes(base64));
const hexToBase64 = hex => bytesToBase64(toBytes.hexToBytes(hex));
// * JSON
// ! WHY: wasm using base64 string for json
// TODO: should refactor
const bytesToJson = bytes => {
  const jsonInStr = bytesToString(bytes);
  return JSON.parse(jsonInStr);
};
const jsonToBytes = json => {
  const jsonInStr = JSON.stringify(json, null, 0);
  return stringToBytes(jsonInStr);
};

exports.base64ToBytes = base64ToBytes;
exports.base64ToHex = base64ToHex;
exports.bytesToBase64 = bytesToBase64;
exports.bytesToJson = bytesToJson;
exports.hexToBase64 = hexToBase64;
exports.jsonToBytes = jsonToBytes;
